File: nim/TMsimulator.nim
nil
type
  symbol* = char

##  The 'symbol' is a char (256)

type
  direction* = enum
    DIR_LEFT, STOP, DIR_RIGHT


##  possible directions
##  struct which contains the result of a transition of the TM

type
  transition_result* {.bycopy.} = object
    control_state*: cint       ##  new state
    write_symbol*: symbol      ##  written symbol


var dir*: direction


##  the state of the TM

type
  turing_machine_state* {.bycopy.} = object
    control_state*: cint       ##  current state
    head_position*: cint       ##  position of the head in the tape
    max_head_position*: cint   ##  number of the most-right visited cell (>=0)
    min_head_position*: cint   ##  number of the most-left visited cell  (<=0)
    tape_r_size*: cint         ##  size of the right tape
    tape_l_size*: cint         ##  size of the left tape
    tape_r*: ptr symbol         ##  right tape (array of symbols)
    tape_l*: ptr symbol         ##  left tape (array of symbols)


##  General data of the TM

type
  turing_machine* {.bycopy.} = object
    number_colors*: cint       ##  number of symbols
    number_states*: cint       ##  number of states
    halting_state*: cint       ##  halting state
    initial_control_state*: cint ##  initial state
    blank_symbol*: symbol      ##  blank symbol
    state*: turing_machine_state
    transition_table*: ptr ptr transition_result ##  REMOVE ????


proc show_data*(m: ptr turing_machine) =
  discard

## ****************************************************************************
##    Executes a step in the given TM
## ***************************************************************************

proc run_step*(m: ptr turing_machine): cint =
  discard

## *****************************************************************
##   Initialization of a Turing Machine
## ****************************************************************

proc init_turing_machine*(states: cint; colors: cint; blank: cint; numberTM: mpz_t): turing_machine =
  discard

proc delete_state*(m: ptr turing_machine) =
  discard

proc outputTM*(m: ptr turing_machine): string =
  discard

proc main*(argn: cint; argv: ptr cstring): cint =
  discard

File: go/TMsimulator.go
package main

/** Rule Components **/
type Symbol byte

type Motion byte

const minSz = 16

type Rule struct {
	State
	Symbol
	Write Symbol
	Motion
	Next State
}

const (
	Left  Motion = 'L'
	Right Motion = 'R'
	Stay  Motion = 'S'
)

func (i *Rule) key() key       { return key{i.State, i.Symbol} }
func (i *Rule) action() action { return action{i.Write, i.Motion, i.Next} }

type key struct {
	State
	Symbol
}

type action struct {
	write Symbol
	Motion
	next State
}

/** Tape Components **/

type Tape struct {
	data      []Symbol
	pos, left int
	blank     Symbol
}
type State string

/** Machine Components **/

type Machine struct {
	tape         *Tape
	start, state State
	transition   map[key]action
	l            func(string, ...interface{}) // XXX
}

// NewTape returns a new tape filled with 'data' and position set to 'start'.
// 'start' does not need to be range, the tape will be extended if required.
func NewTape(blank Symbol, start int, data []Symbol) *Tape {
	t := &Tape{
		data:  data,
		blank: blank,
	}
	if start < 0 {
		t.Left(-start)
	}
	t.Right(start)
	return t
}
func NewMachine(rules []Rule) *Machine {
	m := &Machine{transition: make(map[key]action, len(rules))}
	if len(rules) > 0 {
		m.start = rules[0].State
	}
	for _, r := range rules {
		m.transition[r.key()] = r.action()
	}
	return m
}

//corresponds to init_turing_machine in the TMSimulator.c
func enumerateTM(int states, int colors, mpz_t numberTM) []Rule {
	var table []Rule

}
